Kernel/Graphics/VirtIOGPU/VirtIOGPU.cpp

```diff
@@ -0,0 +1,387 @@
+                dbgln_if(VIRTIO_DEBUG, "VirtIOGPU: VIRGL is not yet supported!");
+            if (is_feature_set(supported_features, VIRTIO_GPU_F_EDID))
+                dbgln_if(VIRTIO_DEBUG, "VirtIOGPU: EDID is not yet supported!");
+            return negotiated;
+        });
+        if (success) {
+            read_config_atomic([&]() {
+                m_num_scanouts = config_read32(*cfg, 0x8);
+            });
+            dbgln_if(VIRTIO_DEBUG, "VirtIOGPU: num_scanouts: {}", m_num_scanouts);
+            success = setup_queues(2); // CONTROLQ + CURSORQ
+        }
+        VERIFY(success);
+        finish_init();
+        Locker locker(m_operation_lock);
+        // 1. Get display information using VIRTIO_GPU_CMD_GET_DISPLAY_INFO
+        query_display_information();
+        // 2. Create BUFFER using VIRTIO_GPU_CMD_RESOURCE_CREATE_2D
+        m_framebuffer_id = create_2d_resource(m_display_info.rect);
+        // 3. Attach backing storage using  VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING
+        // FIXME: We really should be trying to allocate a small amount of pages initially, with ensure_backing_storage increasing the backing memory of the region as needed
+        size_t buffer_length = calculate_framebuffer_size(MAX_VIRTIOGPU_RESOLUTION_WIDTH, MAX_VIRTIOGPU_RESOLUTION_HEIGHT);
+        m_framebuffer = MM.allocate_kernel_region(page_round_up(buffer_length), "VirtGPU FrameBuffer", Region::Access::Read | Region::Access::Write, AllocationStrategy::AllocateNow);
+        ensure_backing_storage(*m_framebuffer, buffer_length, m_framebuffer_id);
+        // 4. Use VIRTIO_GPU_CMD_SET_SCANOUT to link the framebuffer to a display scanout.
+        set_scanout_resource(m_chosen_scanout.value(), m_framebuffer_id, m_display_info.rect);
+        // 5. Render our test pattern
+        draw_ntsc_test_pattern();
+        // 6. Use VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D to update the host resource from guest memory.
+        transfer_framebuffer_data_to_host(m_display_info.rect);
+        // 7. Use VIRTIO_GPU_CMD_RESOURCE_FLUSH to flush the updated resource to the display.
+        flush_displayed_image(m_display_info.rect);
+    } else {
+        VERIFY_NOT_REACHED();
+    }
+}
+
+VirtIOGPU::~VirtIOGPU()
+{
+}
+
+bool VirtIOGPU::handle_device_config_change()
+{
+    return false;
+}
+
+void VirtIOGPU::handle_queue_update(u16 queue_index)
+{
+    dbgln_if(VIRTIO_DEBUG, "VirtIOGPU: Handle queue update");
+    VERIFY(queue_index == CONTROLQ);
+
+    auto& queue = get_queue(CONTROLQ);
+    ScopedSpinLock queue_lock(queue.lock());
+    queue.discard_used_buffers();
+    m_outstanding_request.wake_all();
+}
+
+u32 VirtIOGPU::get_pending_events()
+{
+    return config_read32(*m_device_configuration, DEVICE_EVENTS_READ);
+}
+
+void VirtIOGPU::clear_pending_events(u32 event_bitmask)
+{
+    config_write32(*m_device_configuration, DEVICE_EVENTS_READ, event_bitmask);
+}
+
+void VirtIOGPU::query_display_information()
+{
+    ensure_is_locked();
```

Comment:
Since this function just does a VERIFY call, it might be nicer to just do it inline

Reply:
Will fix in next revision.